companion/src/radiodata.cpp

```diff
@@ -39,61 +39,97 @@ RawSource RawSource::convert(Board::Type before, Board::Type after)
     }
   }
 
-  // No S3, LS and RS on X7 board
-  if (IS_TARANIS_X7(after)) {
-    if (type == SOURCE_TYPE_STICK && index >= 6) {
-      index = 5;
-    }
-  }
-
   // SWI to SWR don't exist on !X9E board
   if (!IS_TARANIS_X9E(after) && IS_TARANIS_X9E(before)) {
     if (type == SOURCE_TYPE_SWITCH && index >= 8) {
       index = index % 8;
     }
   }
 
-  // No SE and SG on X7 board
   if (IS_TARANIS_X7(after)) {
-    if (IS_TARANIS_X9(before) || IS_HORUS(before)) {
-      if (type == SOURCE_TYPE_SWITCH && index >= 6) {
-        index -= 2;
+    // No S3, LS and RS on X7 board
+    if (type == SOURCE_TYPE_STICK && index >= 6) {
+      index = 5;
+    }
+
+    // No SE and SG on X7 board
+    if ((IS_TARANIS_X9(before) || IS_HORUS(before)) && type == SOURCE_TYPE_SWITCH) {
+      if (index == 4 || index == 6) {
+        index = 3;  // SG and SE to SD
+      }
+      else if (index == 5) {
+        index = 4;  // SF to SF
       }
-      else if (type == SOURCE_TYPE_SWITCH && index >= 4) {
-        index -= 1;
+      else if (index == 7) {
+        index = 5;  // SH to SH
       }
     }
   }
 
+  // Compensate for SE and SG on X9/Horus board if converting from X7
+  if ((IS_TARANIS_X9(after) || IS_HORUS(after)) && IS_TARANIS_X7(before) && type == SOURCE_TYPE_SWITCH) {
+    if (index == 4) {
+      index = 5;  // SF to SF
+    }
+    else if (index == 5) {
+      index = 7;  // SH to SH
+    }
+  }
+
   return *this;
 }
 
 RawSwitch RawSwitch::convert(Board::Type before, Board::Type after)
 {
+  if (!index || type != SWITCH_TYPE_SWITCH) {
+    return *this;  // no changes
+  }
+
   // SWI to SWR don't exist on !X9E board
   if (!IS_TARANIS_X9E(after) && IS_TARANIS_X9E(before)) {
-    if (type == SWITCH_TYPE_SWITCH && index >= 24) {
+    if (abs(index) > 24) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Merge this if statement with the enclosing one. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=cpp%3AS1066)